### PR TITLE
JDK-8305248: TableView not rendered correctly after column is made visible if fixed cell size is set

### PR DESCRIPTION
**Pull request by Maran23**
The determined `prefWidth` of a `TableCell` could be `0.0` when a `fixedCellSize` is set.
This happened because the `TableCell` may not have a skin since it was never added to the scene graph yet.

The fix is to make sure we get the `prefWidth` after the `TableCell` was added to the scene graph.
That is also the reason why the problem only happened the first time and never again after (skin is then already created).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8305248](https://bugs.openjdk.org/browse/JDK-8305248): TableView not rendered correctly after column is made visible if fixed cell size is set

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - Committer)
 * [Jose Pereda](https://openjdk.org/census#jpereda) (@jperedadnr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1077/head:pull/1077` \
`$ git checkout pull/1077`

Update a local copy of the PR: \
`$ git checkout pull/1077` \
`$ git pull https://git.openjdk.org/jfx.git pull/1077/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1077`

View PR using the GUI difftool: \
`$ git pr show -t 1077`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1077.diff">https://git.openjdk.org/jfx/pull/1077.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
@andy-goryachev-oracle can you also review this?

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 1]
```diff
@@ -349,12 +349,13 @@ protected ObjectProperty<Node> graphicProperty() {
...
 
-            width = tableCell.prefWidth(height);
-
             if (isVisible) {
                 if (fixedCellSizeEnabled && tableCell.getParent() == null) {
                     getChildren().add(tableCell);
                 }
+                // Note: We have to determine the pref width here because the add operation above may trigger the skin
+                // creation first, which is what makes it possible to get a correct value here in the first place.
+                width = tableCell.prefWidth(height);
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:358]: I wonder if it's safe to move this call (one one after else:430) before the if statement on line 352?

**Review by andy-goryachev-oracle — APPROVED** [review]
thank you for a quick turnaround!
tested with the MonkeyTester with and without fixed cell height, different models, different resize policies.

https://github.com/andy-goryachev-oracle/Test/blob/main/src/goryachev/monkey/MonkeyTesterApp.java

if you want to make one change (assuming if it's safe to do so), i'll re-approve.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 1 above)
Comment (Maran23) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:358]: I'm not sure if I understand correctly: You mean call `tableCell.prefWidth(height)` above?
I chose this way so we always call `tableCell.prefWidth(height)` just once, never twice.

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:358]: actually, please disregard my suggestion:  the first `width = tableCell.prefWidth(height);` happens after adding the `tableCell` to a parent, so it is likely would not be an equivalent change.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 2]
```diff
@@ -349,12 +349,13 @@ protected ObjectProperty<Node> graphicProperty() {
...
             }
 
-            width = tableCell.prefWidth(height);
-
             if (isVisible) {
                 if (fixedCellSizeEnabled && tableCell.getParent() == null) {
                     getChildren().add(tableCell);
                 }
+                // Note: We have to determine the pref width here because the add operation above may trigger the skin
+                // creation first, which is what makes it possible to get a correct value here in the first place.
```
Comment (jperedadnr) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:357]: The comment needs some rewording. As is now, it only explains why the change proposed in this PR (moving the `prefWidth()` call, that is), but then it will be out of context.

Something like:
```
// prefWidth() has to be called only after the tableCell is added to the tableRow, if it wasn't already. 
// Otherwise it might not have its skin yet, and its width value would be 0.
```

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 1 above)
Comment (jperedadnr) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:358]: Alternatively, you could have just kept the first call to `tableCell::prefWidth` as it was, and add a second one only inside the above if expression, right after the tableCell is added to the tableRow. 
I take that, only for such case, there would be two calls instead of one, but it seems to be somehow a cleaner patch and explanation

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableRowSkinTest.java [hunk 3]
```diff
@@ -244,6 +247,47 @@ public void tableRowWithFixedSizeShouldIgnoreVerticalPadding() {
...
+        for (Node cellNode : row.getChildrenUnmodifiable()) {
+            double width = ((Region) cellNode).getWidth();
+            assertNotEquals(0.0, width);
+        }
+    }
+
+    @Test
+    public void testMakeVisibleColumnInvisible() {
+        tableView.setFixedCellSize(24);
+        TableColumn<Person, ?> firstColumn = tableView.getColumns().get(0);
```
Comment (jperedadnr) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableRowSkinTest.java:276]: Might be unnecessary, but you could assert that the first column was visible, just to make sure that the next call to make it invisible has a real change?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableRowSkinTest.java [hunk 4]
```diff
@@ -244,6 +247,47 @@ public void tableRowWithFixedSizeShouldIgnoreVerticalPadding() {
         assertEquals(height, cell.getHeight(), 0);
     }
 
+    /**
+     * When we make an invisible column visible we expect the underlying cells to be visible, e.g. as width > 0.
```
Comment (jperedadnr) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableRowSkinTest.java:251]: This only applies to the case of fixed cell size set (this test will fail if you comment out line 256), so maybe you could clarify the comment?

**Review by jperedadnr — COMMENTED** [review]
The proposed fix works and solves the test case in JDK-8305248. One of the added tests fails without the fix, passes with it (the other one passes in both cases).

I have added a few comments

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 1 above)
Comment (Maran23) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:358]: True and I also thought about it, but I'm still in favor of calling it only as much as needed

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableRowSkinTest.java (hunk 4 above)
Comment (Maran23) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableRowSkinTest.java:251]: Yes! Will do.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java [hunk 5]
```diff
@@ -349,12 +349,13 @@ protected ObjectProperty<Node> graphicProperty() {
                 height = h;
             }
 
-            width = tableCell.prefWidth(height);
-
             if (isVisible) {
                 if (fixedCellSizeEnabled && tableCell.getParent() == null) {
                     getChildren().add(tableCell);
                 }
+                // Note: prefWidth() has to be called only after the tableCell is added to the tableRow, if it wasn't
```
Comment (jperedadnr) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:356]: Minor: there is no explicit rule about limiting line comments to 80 characters, as far as I know, but it might be good to apply it to your comments (also to the test ones)

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java [hunk 6]
```diff
@@ -252,6 +251,53 @@ public void treeTableRowWithFixedCellSizeShouldIgnoreVerticalPadding() {
         assertEquals(height, cell.getHeight(), 0);
     }
 
+    /**
+     * When we set a fixed cell size and make an invisible column visible we expect the underlying cells to be visible,
+     * e.g. width > 0.
+     * See also: JDK-8305248
+     */
+    @Test
+    public void testMakeInvisibleColumnVisible() {
```
Comment (jperedadnr) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:260]: These two tests also pass with/without your patch, but I guess we want to have them to prevent any future issue?

**Review by jperedadnr — COMMENTED** [review]
Looks good, I have just two minor comments.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java (hunk 6 above)
Comment (Maran23) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:260]: Yes. `TreeTableView` does a lot more layout passes then `TableView`. 
I had a look on this in the past and in general the table implementations do too much (re)layouting unfortunately, and that sometimes masks bugs, like this one.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java (hunk 5 above)
Comment (jperedadnr) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableRowSkinBase.java:356]: Ok, right, I missed that.

**Review by jperedadnr — APPROVED** [review]
Looks good to me. The PR fixes the issue, a test fails before it and passes with it, and there are some more tests that, even without failing before this PR, are good to have to prevent future issues.

**Review by andy-goryachev-oracle — APPROVED** [review]
looks good
